### ISSUE TITLE
Question: Rollup value across batches is not rolled up, what am I doing wrong

### ISSUE THREAD

**Issue by attilah** [jamessimone/apex-rollup#376]
I've a rollup setup that `SUM`s a field on `Contact` and rolls up the value to `Account`. Every field has a value of `1` and org default settings for batch size set to `50`. I've created 500 `Contact`s assigned to the same `Account` so the rolled up value should be 500. It is 47 instead (always batch size - 3).

What am I doing wrong? Or is it the expected behavior? because I don't know how any rollup can be done over 100s or 1000s of records.

What I see is that the "prior value" is read out before a given batch starts, and not just before the update and there is no temporary storage where the per batch rollup values would be stored?

**Comment by jamessimone** [comment]
@attilah are you using a calc item where clause? For that few records, you shouldn't even need batching at all - the rollups should be done by a Queuable - but in general cross-batch values are stored using `Database.Stateful`.

**Comment by attilah** [comment]
Nope, it is the simplest rollup you can imagine:

<img width="1312" alt="image" src="https://user-images.githubusercontent.com/230432/196266967-872f9e4c-652c-4ec1-8405-684745663403.png">

Org settings:

<img width="1307" alt="image" src="https://user-images.githubusercontent.com/230432/196267092-58babeed-138c-4392-963f-47f869af1faa.png">

And I used this APEX code to create the 500 test `Contact`s:
```apex
String accountId = '001DJ00000jeaPlYAI';
Integer contactCount = 500;
List<Contact> cc = new List<Contact>();

for (Integer i=0; i<contactCount; i++) {
    Contact c = new Contact();
    c.AccountId = accountId;
    c.Count1__c = 1;
    c.FirstName = 'test';
    c.LastName = 'test';
    
    cc.add(c);
}

insert cc;
```

**Comment by jamessimone** [comment]
Ok let me take a look to see if I can reproduce this issue. I’ll keep you updated.

**Comment by jamessimone** [comment]
@attilah this should be all set in the latest version

**Comment by attilah** [comment]
@jamessimone thanks for looking into it and the quick fix, how does the new parameter affects when we're processing millions of records?
How that new parameter value should be set?

That is our use case, I just used the small batch size for an easier repro.

**Comment by attilah** [comment]
I ran the same test with the default 3000/2000 values for the rollup on 60k records and I see that the conductor is completed, but still having `RollupAsyncProcessor`s running with a bunch of `Maximum stack depth has been reached.` errors and the count is 51996 instead of 60000.

On the 500 test records and small batch sizes the count was correct!

<img width="1095" alt="image" src="https://user-images.githubusercontent.com/230432/196769977-04d1e5cf-51f1-4827-8ae2-93d7df827515.png">

**Comment by jamessimone** [comment]
@attilah which default 3000/2000 values?

I think you might be getting that if the `MaxQueryRows__c` field on Rollup Control is set too low. If that's still at the 25k default, let me know which fields those numbers above were referring to and I will investigate. Also, what kind of environment were you running the above in? Dev Ed orgs have a max queueable re-chaining limit of 5, which could also produce that error - looking at the value you ended up getting rolled up, it's _really_ close to finishing the entire rollup, so I may just need to make a small tweak in a few places once I fully understand your setup.

Thanks for the additional info!

**Comment by attilah** [comment]
@jamessimone here is the screenshot:

<img width="1318" alt="image" src="https://user-images.githubusercontent.com/230432/196815032-d9602a22-12b0-45c7-b71d-a4609a366f5b.png">

I think the new config value did not make it to the layout, so it is on the default 25.000 value.

It does not matter what am I setting for `Should Run As`, both produced the same result, I am running it in a scratch org. I was looking at the code and my suspicion was that perhaps the `getAsyncRollup` has some problems with limit checks or instantiating a `QueueableProcessor` instead of ending up with a Batchable...I can be wrong.

How can a `RollupAsyncProcessor` still started and running while the conductor is in Completed status? It does not wait on the workers to finish fully? or it is just a side effect that we see because of the stack depth issues?

**Comment by jamessimone** [comment]
I believe scratch orgs have the same limitation as far as chaining goes. I’ll add the new field to the layout in the next version. Have you attempted to use the Rollup logging at all to see where and how the deferral that ultimately lead to the maximum stack depth error is occurring? That may provide some further clues.

**Comment by jamessimone** [comment]
@attilah I've managed to repro the issue you're having, which is specifically related to a single parent record having more than the max allowable number of queryable children. I'm working on a fix for this and hope to have a release ready with it in the next few days!

**Comment by attilah** [comment]
That's awesome news, thanks for looking into it @jamessimone, I hope this fix will benefit others as well!

**Comment by jamessimone** [comment]
@attilah most definitely. I apologize for the delay on this fix. I got my failing test for this scenario passing and also did some integration testing for a parent with 62,001 children. I am going to see if I can work in some other optimizations to speed up how long it takes for batch calculations where this kind of lookup skew has occurred. It’s possible those optimizations may need to be deferred to a future release - in any event, I’ll be trying to cut the new release for Apex Rollup in the next day or two with at least the initial fix in place

**Comment by attilah** [comment]
@jamessimone I appreciate the effort you make, don't apologize please, it is a valuable OSS project, everyone should know what they signing up for! Take your time, once it is out we'll provide feedback on the test's result, but as you have a repro it will "just work". Thanks for the update!

**Comment by jamessimone** [comment]
@attilah you should now be good to go - I will work on optimizing the batch speed in the next release, but the calculations should all roll up correctly now.

**Comment by attilah** [comment]
Hey @jamessimone I did some testing and the stateful modification did not work as expected based on my testing.

What I am seeing is that only `Batch Chunk Size` records are being rolled up correctly.

I've tested a different scenario though.

- Created a `Property` object with a `PMCount` field
- Created a `PrintMail` custom object with lookup field to `Property`
- Created 70,500 `PrintMail` records, and 14,100 `Property` records. Each `Property` has 5 `PrintMail` objects assigned through the lookup field  

Here is my test script I ran several times for quick and easy test data:
```
Integer iteration = 0; // this is what I incremented to get more records
Integer propertyCount = 1000;
Integer propertyOffset = propertyCount * iteration;
Integer pmCount = 5;
String pad = '0000000000';

List<SObject> objects = new List<SObject>();

for (Integer i=0; i<propertyCount; i++) {
    Property__c property = new Property__c();
    Integer nameValue = propertyOffset + i;

    property.Name = String.valueOf(nameValue).leftPad(8, '0');

    objects.add(property);
}

insert objects;

List<PrintMail__c> pm = new List<PrintMail__c>();

for (Integer i=0; i<propertyCount; i++) {
    for (Integer j=0; j<pmCount; j++) {
        PrintMail__c printMail = new PrintMail__c();
        
        printMail.Property__c = objects.get(i).Id;

        pm.add(printMail);
    }
}

insert pm;
```

This is the rollup config:
<img width="1320" alt="image" src="https://user-images.githubusercontent.com/230432/199844484-caf73414-74cf-48fa-b169-08c000881d4e.png">

This is the org default record:
<img width="1319" alt="image" src="https://user-images.githubusercontent.com/230432/199844799-e88b1aff-fbd7-4896-9d65-f6d6a7d0f7dd.png">

My suspiciton was around the stateful and `CalcItemBag` thingy so:
- I tried to remove the `final` keyword from `statefulLookupToCalcItems` as it was recommended by few to work better with stateful batch jobs, did not help as probably in the background a in memory binary serialization of the object happens in JVM world...
- I tried to make `statefulLookupToCalcItems` static and in that case the count was CORRECT.

I hope this gives some help to troubleshoot this.

**Comment by jamessimone** [comment]
I will take a look, static properties aren’t Stateful but it’s odd that preventing that variable from being Stateful caused this to work. Also, final merely prevents an object from being reinitialized - it has no other effects so I’m not surprised changing that didn’t do anything.

**Comment by attilah** [comment]
Skipping stateful with the static modification was intentional to see if I'm right that the problem lies around that stateful field variable ;-)

**Comment by jamessimone** [comment]
Hey, if it works and everything else works, that would also serve to speed things up quite nicely. I will just have to test to ensure!

**Comment by attilah** [comment]
I don't have in-depth knowledge about how would it be affected by running multiple rollups over different objects, etc...perhaps having it static but cleaning up between the different rollups could help....you are the one who could possibly know it the best!

**Comment by jamessimone** [comment]
That change won't work, unfortunately. The map _needs_ to be an instance property and stateful in order to track changes to parents whose children items exceed the count of each batch chunk. I have managed to reproduce this issue, though, so I should be able to get a fix going in the near future.

**Comment by attilah** [comment]
Yes I agree, it was good "test fix" for my suspicion, but not THE fix. Glad you were able to repro it, will wait for the fix, LMK if I can be any help.

**Comment by jamessimone** [comment]
@attilah still working on some edge cases for this one, but wanted to release my progress so far with #381

**Comment by attilah** [comment]
Thanks @jamessimone we did some initial testing and the rollup numbers were matching what we get from the reports, awesome progress, thanks for making the effort!

**Comment by jamessimone** [comment]
excellent!

**Comment by attilah** [comment]
Hey @jamessimone, further testing revealed that the latest version does not work as expected. We have 2 SUM rollups that we checked over millions of parent records, and child records, but we've `where` clauses on the lookup records, which reduces that count significantly in Rollup 1.

I thought that somehow in case of large record counts some info is lost hence the problem, but does not look like it.

In case of Rollup 1, the where clause on child records filters down the records to 3604, the rolled up SUM value is not correct, we get all the parent record rollup fields zeroed out, but count values are put only in few ~2500ish. Always the same value on multiple runs.

In case of Rollup 2, the where clause filtering down the child records to 1.9M, and it completes, but the result is the same.

**Comment by jamessimone** [comment]
I’ll have to take a look. Where clauses are evaluated in memory, due to all records needing to be pulled back prior to processing, but I think the issue is still surrounding the Stateful batch’s tracking of all children for any given parent. I haven’t been able to fully track what the issue is for these cases. One thing that would be interesting would be to start the full recalculation for just a single one of these rollups using the form version of the full recalc app with the where clause written out, as a singular rollup being done has some behavior specific to it that doesn’t hold true for multiple rollups being performed all at once for a full recalc.

**Comment by jamessimone** [comment]
Making progress on this! I’m hopeful to finish the work up over the next few days

**Comment by jamessimone** [comment]
@attilah wrapping up the release for this tomorrow. I've done extensive testing on this now - everything looks good on my end!

**Comment by attilah** [comment]
@jamessimone thanks for all the effort, can't wait to test it after the release has happened!

**Comment by jamessimone** [comment]
@attilah this is now generally available! I'm also planning to add a new Wiki page about Large Data Volume in general - since updates to parent records will be delayed for as long as it takes to recalculate fully. Let me know if you have any issues!
